src/library/scala/runtime/ClassValueCompat.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Scala (https://www.scala-lang.org)
+ *
+ * Copyright EPFL and Lightbend, Inc.
+ *
+ * Licensed under Apache License 2.0
+ * (http://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package scala.runtime;
+
+public abstract class ClassValueCompat<T> {
```

Comment:
We cannot add a public class in a Scala 2.13.x release due to forwards binary compatibility.

I think the only workaround here is to write the class in Scala (not Java) and mark it `private[scala]`. This requires a [MiMa filter](https://github.com/scala/scala/blob/2.13.x/project/MimaFilters.scala); running `library/mimaReportBinaryIssues` in sbt will tell you what to put in there.

I see that `ClassValueCompat` is used in the Java class `ModuleSerializationProxy` and we're using `CompileOrder.JavaThenScala`, so that's a problem. I guess we can create a private copy of `ClassValueCompat` within `ModuleSerializationProxy`.

Reply:
> We cannot add a public class in a Scala 2.13.x release due to forwards binary compatibility.

Does this restriction applies to classes in scala.runtime too? /cc @sjrd

Reply:
I think so. `scala.reflect.internal` is excluded, but that's only in the "reflect" module, we need the class in the "library" module.

Reply:
> 
> 
> > We cannot add a public class in a Scala 2.13.x release due to forwards binary compatibility.
> 
> Does this restriction applies to classes in scala.runtime too? /cc @sjrd

The thing is that this class shouldn't be in `scala.runtime`. `scala.runtime` is supposed to be for code called by the codegen. Not for code called from library code. Code called from library code should be in the library, next to what calls them. There is no reason to put this class in `scala.runtime`.

Reply:
I'm not sure we need to be so strict about this, there's other stuff in `scala.runtime` too (RichInt).

Reply:
> 
> 
> I'm not sure we need to be so strict about this, there's other stuff in `scala.runtime` too (RichInt).

Yes, and it's not great that they are there, as it muddies the waters between what's public API and what isn't. I had a WiP PR to move them out, but I didn't finish it in time for 2.13.0.

Reply:
Relocated `ClassValueCompat` to be under `scala.compat`, rewritten in Scala and applied the suggestions above mentioned in the comments, thanks for the directions

Reply:
`scala.compat` currently only has a single deprecated class in it. So that package was well on its way to being eliminated entirely, but this PR would bring it back to life.

I might be more comfortable just putting the new class in `scala.runtime`, since doing it that way is the status quo. Then in the future if someone wants to submit a PR like @sjrd's that reorganizes and/or splits `scala.runtime` according to some guiding principle, then great.

@sjrd you objected to using `scala.runtime`, but you didn't (?) propose a better location. wdyt of the choice of `scala.compat`?

Reply:
My suggestion is to put in whatever actual package it is used, as package-private to that package.

Reply:
It's currently being utilized for `scala.runtime` and `scala.reflect`, will there be a way to declare as package-private of two packages?

Reply:
`private[scala]` (the mutual super-package of `scala.runtime` and `scala.reflect`), and probably in the `scala` package. there's some argument to having a `scala.internal` package for this and the sorts of things @sjrd mentioned that are internal but not generated by the compiler, but I'm not sure if it's worth investing in creating that package outside of a major release or not.

Reply:
`ClassValueCompat` Relocated as `private[scala]` under `scala` package as suggested